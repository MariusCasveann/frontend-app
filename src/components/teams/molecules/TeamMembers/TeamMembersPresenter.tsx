import React, { useState } from 'react';
import { Divider, Table, Avatar, Button } from 'antd';
import { ColumnProps } from 'antd/lib/table';
import LoadingOrErrorPresenter from '../../../common/organisms/LoadingOrError/LoadingOrErrorPresenter';
import { Card } from '../../../common/molecules/CardLayout/CardLayout';
import {
    TeamByProductId_teamByProductId,
    TeamByProductId_teamByProductId_teamMembers
} from './graphql/__generated__/TeamByProductId';
import {
    QueryCurrentUser_currentUser,
    QueryCurrentUser_currentUser_roles
} from '../../../authentication/molecules/UserProfile/graphql/__generated__/QueryCurrentUser';
import { QueryAllProducts_allProducts } from '../../../dashboard/atoms/ProductSelection/graphql/__generated__/QueryAllProducts';
import TeamProductOwnerPresenter from './TeamProductOwnerPresenter';
import { checkIfTeamHasData, checkIfMemberIsAdmin, checkIfCurrentUserIsAdmin } from '../../helpers/teamHelper';
import AdminRoleButton from '../../atoms/AdminRoleButton';
import ConfirmModal from '../../../common/molecules/Modals/ConfirmModal';
import '../../../common/molecules/CardLayout/CardLayout.css';
import './Team.css';
import CreateTeam from '../../../createTeam/CreateTeam';
import EditProductOwner from '../ProductOwner/EditProductOwner';
import AddTeamMembers from '../AddTeamMembers/AddTeamMembers';
import { History, LocationState } from 'history';

export interface TeamMemberProps {
    teamByProductId: TeamByProductId_teamByProductId;
    currentUser: QueryCurrentUser_currentUser | null;
    loading: boolean;
    error: string;
    selectedProduct: QueryAllProducts_allProducts;
    onAssign: (productId: string, userId: string) => void;
    onRevoke: (productId: string, userId: string) => void;
    onDelete: (productId: string, team: TeamByProductId_teamByProductId, userId: string) => void;
    history: History<LocationState>;
}

export const getColumns = (
    onAssign: (productId: string, userId: string) => void,
    onRevoke: (productId: string, userId: string) => void,
    selectedProduct: QueryAllProducts_allProducts,
    setUserId: (userId: string) => void,
    setVisible: (visbile: boolean) => void,
    isCurrentUserAdmin?: QueryCurrentUser_currentUser_roles | null
) => {
    const columns: Array<ColumnProps<TeamByProductId_teamByProductId_teamMembers | null>> = [
        {
            render: () => <Avatar size="small" icon="user" />,
            width: 20,
            key: 'avatar'
        },
        {
            title: 'Name',
            render: name => `${name.firstName} ${name.lastName}`,
            key: 'name'
        },
        {
            title: 'Email',
            dataIndex: 'email',
            key: 'email'
        },
        {
            title: 'Action',
            key: 'action',
            render: item => (
                <div className="member-action-button">
                    {isCurrentUserAdmin ? (
                        <>
                            <AdminRoleButton
                                onUpdateRole={() => {
                                    const admin = checkIfMemberIsAdmin(item, selectedProduct);
                                    admin
                                        ? onRevoke(selectedProduct.id, item.id)
                                        : onAssign(selectedProduct.id, item.id);
                                }}
                                isTeamMemberAdmin={checkIfMemberIsAdmin(item, selectedProduct)}
                            />
                            <Divider type="vertical" />
                        </>
                    ) : (
                        undefined
                    )}
                    <Button
                        type="link"
                        size="small"
                        onClick={() => {
                            setUserId(item.id);
                            setVisible(true);
                        }}
                    >
                        Delete
                    </Button>
                </div>
            )
        }
    ];
    return columns;
};

export default ({
    teamByProductId,
    loading,
    error,
    currentUser,
    selectedProduct,
    onAssign,
    onRevoke,
    onDelete,
    history
}: TeamMemberProps) => {
    const isCurrentUserAdmin = checkIfCurrentUserIsAdmin(currentUser, selectedProduct);
    const [visible, setVisible] = useState(false);
    const [userId, setUserId] = useState('');
    const [changeProductOwner, setChangeProductOwner] = useState<boolean>(false);
    const [addTeamMembers, setAddTeamMembers] = useState<boolean>(false);

    if (loading || error) {
        return (
            <Card width={100}>
                <LoadingOrErrorPresenter error={error} loading={loading} />
            </Card>
        );
    }
    if (checkIfTeamHasData(teamByProductId)) {
        return (
            <>
                <ConfirmModal
                    title="Delete"
                    visible={visible}
                    onCancel={() => setVisible(false)}
                    onOk={async () => {
                        await onDelete(selectedProduct.id, teamByProductId, userId);

                        setVisible(false);
                    }}
                    text="Do you want to delete this team member"
                />
                <Card title="Team members" width={100}>
                    <EditProductOwner
                        selectedProduct={selectedProduct}
                        teamByProductId={teamByProductId}
                        changeProductOwner={changeProductOwner}
                        setChangeProductOwner={setChangeProductOwner}
                    />
                    <AddTeamMembers
                        selectedProduct={selectedProduct}
                        teamByProductId={teamByProductId}
                        addTeamMembers={addTeamMembers}
                        setAddTeamMembers={setAddTeamMembers}
                    />
                    <TeamProductOwnerPresenter
                        teamByProductId={teamByProductId}
                        setChangeProductOwner={setChangeProductOwner}
                        setAddTeamMembers={setAddTeamMembers}
                    />
                    <Table
                        rowKey={record => (record && record.lastName) || ''}
                        columns={getColumns(
                            onAssign,
                            onRevoke,
                            selectedProduct,
                            setUserId,
                            setVisible,
                            isCurrentUserAdmin
                        )}
                        dataSource={(teamByProductId && teamByProductId.teamMembers) || []}
                        pagination={false}
                    />
                </Card>
            </>
        );
    }
    return <CreateTeam history={history} />;
};
