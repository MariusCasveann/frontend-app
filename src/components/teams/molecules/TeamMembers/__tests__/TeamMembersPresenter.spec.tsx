import * as React from 'react';
import renderer from 'react-test-renderer';
jest.mock('antd', () => ({ Table: 'Table', Divider: 'Divider', Avatar: 'Avatar', Button: 'Button' }));
jest.mock('antd/lib/table', () => ({ ColumnProps: 'ColumnProps' }));
jest.mock('../../../../common/organisms/LoadingOrError/LoadingOrErrorPresenter', () => 'LoadingOrErrorPresenter');
jest.mock('../../../../common/molecules/CardLayout/CardLayout', () => ({ Card: 'Card' }));
jest.mock('./../TeamProductOwnerPresenter', () => 'TeamProductOwnerPresenter');
jest.mock('../../../atoms/AdminRoleButton', () => 'AdminRoleButtonu');
jest.mock('../../../../common/molecules/Modals/ConfirmModal', () => 'ConfirmModal');
jest.mock('../../../../createTeam/CreateTeam', () => 'CreateTeam');
jest.mock('../../ProductOwner/EditProductOwner', () => 'EditProductOwner');
jest.mock('../../AddTeamMembers/AddTeamMembers', () => 'AddTeamMembers');

import {
    mockSelectedProduct,
    mockErrorMessage,
    mockTeamByProductId,
    mockCurrentUser,
    mockEmptyTeam,
    mockCurrenUserRole,
    mockFunction,
    mockHistory
} from '../../../../common/mock/MockData';
import TeamMembersPresenter from '../TeamMembersPresenter';
import { getColumns } from '../TeamMembersPresenter';

describe('TeamMembersPresenter', () => {
    describe('when there is value', () => {
        it('renders a Card with teammembers', () => {
            const wrapper = renderer
                .create(
                    <TeamMembersPresenter
                        onAssign={mockFunction}
                        onDelete={mockFunction}
                        onRevoke={mockFunction}
                        teamByProductId={mockTeamByProductId}
                        currentUser={mockCurrentUser}
                        history={{ ...mockHistory, action: 'PUSH' }}
                        loading={false}
                        error={''}
                        selectedProduct={mockSelectedProduct}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });
    describe('when there is an error', () => {
        it('renders LoadingOrError', () => {
            const wrapper = renderer
                .create(
                    <TeamMembersPresenter
                        onAssign={mockFunction}
                        onDelete={mockFunction}
                        onRevoke={mockFunction}
                        teamByProductId={mockTeamByProductId}
                        currentUser={mockCurrentUser}
                        history={{ ...mockHistory, action: 'PUSH' }}
                        loading={false}
                        error={mockErrorMessage}
                        selectedProduct={mockSelectedProduct}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });
    describe('when there is loading', () => {
        it('renders LoadingOrError', () => {
            const wrapper = renderer
                .create(
                    <TeamMembersPresenter
                        onAssign={mockFunction}
                        onDelete={mockFunction}
                        onRevoke={mockFunction}
                        teamByProductId={mockTeamByProductId}
                        currentUser={mockCurrentUser}
                        history={{ ...mockHistory, action: 'PUSH' }}
                        loading={true}
                        error={''}
                        selectedProduct={mockSelectedProduct}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });
    describe('when there is no data', () => {
        it('renders no data', () => {
            const wrapper = renderer
                .create(
                    <TeamMembersPresenter
                        onAssign={mockFunction}
                        onDelete={mockFunction}
                        onRevoke={mockFunction}
                        teamByProductId={mockEmptyTeam}
                        currentUser={mockCurrentUser}
                        loading={false}
                        history={{ ...mockHistory, action: 'PUSH' }}
                        error={''}
                        selectedProduct={mockSelectedProduct}
                    />
                )
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });
});
describe('getColumns method', () => {
    it('should return 4', () => {
        const columnLength = getColumns(
            mockFunction,
            mockFunction,
            mockSelectedProduct,
            mockFunction,
            mockFunction,
            mockCurrenUserRole
        );
        expect(columnLength).toHaveLength(4);
    });
});
