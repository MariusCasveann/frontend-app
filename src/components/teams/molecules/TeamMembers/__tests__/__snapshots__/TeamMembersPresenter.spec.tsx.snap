// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamMembersPresenter when there is an error renders LoadingOrError 1`] = `
<Card
  width={100}
>
  <LoadingOrErrorPresenter
    error="Error message"
    loading={false}
  />
</Card>
`;

exports[`TeamMembersPresenter when there is loading renders LoadingOrError 1`] = `
<Card
  width={100}
>
  <LoadingOrErrorPresenter
    error=""
    loading={true}
  />
</Card>
`;

exports[`TeamMembersPresenter when there is no data renders no data 1`] = `
<CreateTeam
  history={
    Object {
      "action": "PUSH",
      "block": [MockFunction],
      "createHref": [MockFunction],
      "go": [MockFunction],
      "goBack": [MockFunction],
      "goForward": [MockFunction],
      "length": 7,
      "listen": [MockFunction],
      "location": Object {
        "hash": "",
        "pathname": "/path",
        "search": "",
        "state": Object {},
      },
      "push": [MockFunction],
      "replace": [MockFunction],
    }
  }
/>
`;

exports[`TeamMembersPresenter when there is value renders a Card with teammembers 1`] = `
Array [
  <ConfirmModal
    onCancel={[Function]}
    onOk={[Function]}
    text="Do you want to delete this team member"
    title="Delete"
    visible={false}
  />,
  <Card
    title="Team members"
    width={100}
  >
    <EditProductOwner
      changeProductOwner={false}
      selectedProduct={
        Object {
          "__typename": "LightProductDTO",
          "id": 50051,
          "logo": null,
          "name": "a",
        }
      }
      setChangeProductOwner={[Function]}
      teamByProductId={
        Object {
          "__typename": "Team",
          "id": 2355,
          "productOwner": Object {
            "__typename": "User",
            "firstName": "TestUser",
            "id": 11,
            "lastName": "Test",
          },
          "teamMembers": Array [
            Object {
              "__typename": "User",
              "email": "test1@mediamarktsaturn.com",
              "firstName": "Test",
              "id": "1234",
              "lastName": "Test2",
              "roles": Array [
                Object {
                  "__typename": "Role",
                  "name": "WRITE_ALL",
                },
                Object {
                  "__typename": "Role",
                  "name": "a-admin",
                },
              ],
            },
            Object {
              "__typename": "User",
              "email": "test2@mediamarktsaturn.com",
              "firstName": "Test2",
              "id": "2234",
              "lastName": "Test2",
              "roles": Array [
                Object {
                  "__typename": "Role",
                  "name": "WRITE_ALL",
                },
                Object {
                  "__typename": "Role",
                  "name": "a-admin",
                },
                Object {
                  "__typename": "Role",
                  "name": "test-admin",
                },
              ],
            },
          ],
        }
      }
    />
    <AddTeamMembers
      addTeamMembers={false}
      selectedProduct={
        Object {
          "__typename": "LightProductDTO",
          "id": 50051,
          "logo": null,
          "name": "a",
        }
      }
      setAddTeamMembers={[Function]}
      teamByProductId={
        Object {
          "__typename": "Team",
          "id": 2355,
          "productOwner": Object {
            "__typename": "User",
            "firstName": "TestUser",
            "id": 11,
            "lastName": "Test",
          },
          "teamMembers": Array [
            Object {
              "__typename": "User",
              "email": "test1@mediamarktsaturn.com",
              "firstName": "Test",
              "id": "1234",
              "lastName": "Test2",
              "roles": Array [
                Object {
                  "__typename": "Role",
                  "name": "WRITE_ALL",
                },
                Object {
                  "__typename": "Role",
                  "name": "a-admin",
                },
              ],
            },
            Object {
              "__typename": "User",
              "email": "test2@mediamarktsaturn.com",
              "firstName": "Test2",
              "id": "2234",
              "lastName": "Test2",
              "roles": Array [
                Object {
                  "__typename": "Role",
                  "name": "WRITE_ALL",
                },
                Object {
                  "__typename": "Role",
                  "name": "a-admin",
                },
                Object {
                  "__typename": "Role",
                  "name": "test-admin",
                },
              ],
            },
          ],
        }
      }
    />
    <TeamProductOwnerPresenter
      setAddTeamMembers={[Function]}
      setChangeProductOwner={[Function]}
      teamByProductId={
        Object {
          "__typename": "Team",
          "id": 2355,
          "productOwner": Object {
            "__typename": "User",
            "firstName": "TestUser",
            "id": 11,
            "lastName": "Test",
          },
          "teamMembers": Array [
            Object {
              "__typename": "User",
              "email": "test1@mediamarktsaturn.com",
              "firstName": "Test",
              "id": "1234",
              "lastName": "Test2",
              "roles": Array [
                Object {
                  "__typename": "Role",
                  "name": "WRITE_ALL",
                },
                Object {
                  "__typename": "Role",
                  "name": "a-admin",
                },
              ],
            },
            Object {
              "__typename": "User",
              "email": "test2@mediamarktsaturn.com",
              "firstName": "Test2",
              "id": "2234",
              "lastName": "Test2",
              "roles": Array [
                Object {
                  "__typename": "Role",
                  "name": "WRITE_ALL",
                },
                Object {
                  "__typename": "Role",
                  "name": "a-admin",
                },
                Object {
                  "__typename": "Role",
                  "name": "test-admin",
                },
              ],
            },
          ],
        }
      }
    />
    <Table
      columns={
        Array [
          Object {
            "key": "avatar",
            "render": [Function],
            "width": 20,
          },
          Object {
            "key": "name",
            "render": [Function],
            "title": "Name",
          },
          Object {
            "dataIndex": "email",
            "key": "email",
            "title": "Email",
          },
          Object {
            "key": "action",
            "render": [Function],
            "title": "Action",
          },
        ]
      }
      dataSource={
        Array [
          Object {
            "__typename": "User",
            "email": "test1@mediamarktsaturn.com",
            "firstName": "Test",
            "id": "1234",
            "lastName": "Test2",
            "roles": Array [
              Object {
                "__typename": "Role",
                "name": "WRITE_ALL",
              },
              Object {
                "__typename": "Role",
                "name": "a-admin",
              },
            ],
          },
          Object {
            "__typename": "User",
            "email": "test2@mediamarktsaturn.com",
            "firstName": "Test2",
            "id": "2234",
            "lastName": "Test2",
            "roles": Array [
              Object {
                "__typename": "Role",
                "name": "WRITE_ALL",
              },
              Object {
                "__typename": "Role",
                "name": "a-admin",
              },
              Object {
                "__typename": "Role",
                "name": "test-admin",
              },
            ],
          },
        ]
      }
      pagination={false}
      rowKey={[Function]}
    />
  </Card>,
]
`;
